### DISCUSSION TITLE
Delete fetch requests freezing

### DISCUSSION THREAD

**Discussion by jzaplet** [pocketbase/pocketbase#1575] [Q&A]
Hi, I have problem with `DELETE` request in JS-SDK (I tried multiple browsers, anonymous windows) and still don't know where is a problem with `await pb.collection('any_collection_name').delete('xyz')`

`POST` and `GET` requests works absolutely fine, but `DELETE` requests does not return any response - just stuck in pending status.

Backend data are proceeded, but no response returned. You can see it in this video [https://youtu.be/bV3_u2lEGSw](https://youtu.be/bV3_u2lEGSw).

Any idea where is a problem?

**Comment by ganigeorgiev** [comment]
What version of PocketBase do you use?
If you are using older than v0.10.4, could you try with the latest [v0.11.1](https://github.com/pocketbase/pocketbase/releases/tag/v0.11.1)?

Does it ever complete if you leave it to run?

Do you have a lot of relational data?
_Cascade delete/update could be slow in large datasets because currently we don't have indexes for the `relation` fields by default (it will be improved in the near future)._

**Reply by ganigeorgiev** [reply]
If after updating PocketBase, you are still experiencing the issue, could you export your collections schema (Admin UI > Settings > Export collections) and attach it here as json or send it to "support [at] pocketbase.io"?

**Reply by jzaplet** [reply]
Thank you for blazing fast response.

I build application in golang version `1.19`

```
// main.go
package main

import (
    "log"

    "github.com/pocketbase/pocketbase"
)

func main() {
    app := pocketbase.New()

    if err := app.Start(); err != nil {
        log.Fatal(err)
    }
}
```

Pocketbase version is `v0.11.1` and JS-SDK is `0.10.0`.

Same problem occurred when i try native JS `fetch()` function.

My database contains maximum 300 rows in total.

Relations does not affect result because I also tried to remove rows without any relations.

Fetch status after 15 minutes is still `PENDING` 😄  (until manually shutdown server).

Exported collections:
[pb_schema.json.zip](https://github.com/pocketbase/pocketbase/files/10395388/pb_schema.json.zip)

**Reply by ganigeorgiev** [reply]
@jzaplet I'm not able to reproduce the issue. I've tried to create 2k contacts assoiated to 2k companies and 2k invoices each with 500 invoice items. I've even tried disabling the required constraint to allow propagating further the cascade operations and everything seems to work fine when I try to delete a company.

Does the delete stuck if you try to delete the entry from the Admin UI or via a rest client (curl, postman, insomnia, etc.)?

Do you use the realtime api?

Do you have any custom Go code that could interfere with the delete action?

Could you try with the default prebuilt executables?

**Reply by jzaplet** [reply]
@ganigeorgiev 
I downloaded newest executable prebuild for Mac os ARM64 and AMD64 and tested both of them (My working machine is Macbook pro M1).

- Yes I have same problem in Admin UI & in Postman
- I don't use realtime API
- I don't have custom Go code
- I tried default executable (0.11.1) and got same results.

Which browser do you use?

![Screenshot 2023-01-11 at 21 20 46](https://user-images.githubusercontent.com/4992367/211909587-7e832a44-49e1-4276-bda5-c2dec29eb5c9.png)
![Screenshot 2023-01-11 at 21 20 34](https://user-images.githubusercontent.com/4992367/211909599-6ed5ad07-819d-467c-83cf-117d87de77a9.png)

**Reply by ganigeorgiev** [reply]
> Which browser do you use?

I'm using Chrome, but I don't think this is browser related issue.

Something in the Go code is causing the server to hang but I'm not sure what yet.
I'll continue investigating it.

**Reply by jzaplet** [reply]
Ok, I'm trying too.
I sent e-mail to you. This e-mail contains all project data.
Please check your Junk folder. My e-mail is "jz [at] strategio.dev"

**Reply by ganigeorgiev** [reply]
Thank you. I've already found 1 issue related to eagerly closed channel and will try to provide a fix tomorrow.

**Reply by jzaplet** [reply]
Wow, thank you very much! I'll be waiting :-)

**Reply by ganigeorgiev** [reply]
@jzaplet There are 2 issues:

1. When batch deleting records we try to exit early, but we are not terminating the fired go routines which ended up sending to a closed channel (the channel also shouldn't be closed by the receiver, but that's another thing that I'll have to refactor). This in rare situations could result in a panic if some constraint (eg. nonempty) fail on cascade delete/update (_but this is not the reason for the hang_).

2. After successful record delete we are triggering a delete for all files of the cascade deleted records. And this is actually the reason for the hang because in case of incorrect S3 setup, it will wait indefinitely until the other side return a response. You can verify it by going to "Admin UI > Settings > File storage" and probably will see a "Failed to establish S3 connection" error. **If you provide valid S3 settings or temporary disable the S3 storage, you'll find that the delete will complete successfully.**

I'll try to provide fixes (or at least some safe guards) for both issues by the end of the week and will update the integration tests accordingly.

**Reply by jzaplet** [reply]
@ganigeorgiev
Thank you for the prompt investigation. I understand where the problem is. In my case, all I had to do was turn off the S3 settings and everything was working fine.
